diff -uprNw -X linux-4.1.vanilla/Documentation/dontdiff linux-4.1.vanilla/arch/x86/kernel/smpboot.c linux-4.1.missing-sched-domains/arch/x86/kernel/smpboot.c
--- linux-4.1.vanilla/arch/x86/kernel/smpboot.c	2015-06-21 22:05:43.000000000 -0700
+++ linux-4.1.missing-sched-domains/arch/x86/kernel/smpboot.c	2015-11-05 00:54:31.329592670 -0800
@@ -445,7 +445,12 @@ void set_cpu_sibling_map(int cpu)
 			} else if (i != cpu && !c->booted_cores)
 				c->booted_cores = cpu_data(i).booted_cores;
 		}
+#ifndef CONFIG_NUMA
 		if (match_die(c, o) && !topology_same_node(c, o))
+#else
+        if (match_die(c, o) && !topology_same_node(c, o)
+            && sched_max_numa_distance == -1)
+#endif
 			primarily_use_numa_for_topology();
 	}
 }
diff -uprNw -X linux-4.1.vanilla/Documentation/dontdiff linux-4.1.vanilla/kernel/sched/core.c linux-4.1.missing-sched-domains/kernel/sched/core.c
--- linux-4.1.vanilla/kernel/sched/core.c	2015-06-21 22:05:43.000000000 -0700
+++ linux-4.1.missing-sched-domains/kernel/sched/core.c	2015-11-05 01:08:25.009554228 -0800
@@ -6130,7 +6130,7 @@ static void claim_allocations(int cpu, s
 static int sched_domains_numa_levels;
 enum numa_topology_type sched_numa_topology_type;
 static int *sched_domains_numa_distance;
-int sched_max_numa_distance;
+int sched_max_numa_distance = -1;
 static struct cpumask ***sched_domains_numa_masks;
 static int sched_domains_curr_level;
 #endif
